File: 257-binary-tree-paths/binary-tree-paths.py
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        """
                def helper(root,path):
            if root==None:
                return ""
            
            if root.left==None and root.right==None:
                res.append(path+str(root.val))
                return
            helper(root.left,path+str(root.val)+'->')
            helper(root.right,path+str(root.val)+'->')
        #------------------
        res=[]
        helper(root,"") # subset=[]
        return res"""
        def helper(root,subset):
            if root.left==None and root.right==None:
                res.append("->".join(subset))
                return

            if root.left!=None:
                subset.append(str(root.left.val))
                helper(root.left,subset)
                subset.pop()
            
            if root.right!=None:
                subset.append(str(root.right.val))
                helper(root.right,subset)
                subset.pop()
        #------------------------
        res=[]
        helper(root,[str(root.val)])
        return res

File: 257-binary-tree-paths/binary-tree-paths.java

class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> res=new ArrayList<>();
        backtrack(root,"",res); //path=""
        return res;
    }
    private void backtrack(TreeNode root, String path,List<String> res){
        if(root==null){
            return;
        }
        if(root.left==null && root.right==null){
            res.add(path+String.valueOf(root.val));
        }
        backtrack(root.left,path+String.valueOf(root.val)+"->",res);
        backtrack(root.right,path+String.valueOf(root.val)+"->",res);
    }
}